src/str.rs

```diff
@@ -433,186 +445,229 @@ pub(crate) fn parse_str_radix_n(str: &str, radix: u32) -> Result<Decimal, crate:
+                b'.' => {
+                    if passed_decimal_point {
+                        return Err(Error::from("Invalid decimal: two decimal points"));
+                    }
+                    passed_decimal_point = true;
                 }
-                digits_before_dot = coeff.len() as i32;
-                offset += 1;
-                len -= 1;
-            }
-            b'_' => {
-                // Must start with a number...
-                if coeff.is_empty() {
-                    return Err(Error::from("Invalid decimal: must start lead with a number"));
+                b'_' => {
+                    // Must start with a number...
+                    if !has_digit {
+                        return Err(Error::from("Invalid decimal: must start lead with a number"));
+                    }
                 }
-                offset += 1;
-                len -= 1;
+                _ => return Err(Error::from("Invalid decimal: unknown character")),
             }
-            _ => return Err(Error::from("Invalid decimal: unknown character")),
+            bytes = &bytes[1..];
+        } else {
+            break;
         }
     }
 
-    // If we exited before the end of the string then do some rounding if necessary
-    if maybe_round && offset < bytes.len() {
-        let next_byte = bytes[offset];
-        let digit = match next_byte {
-            b'0'..=b'9' => {
-                if next_byte > max_n {
-                    return Err(Error::from("Invalid decimal: invalid character"));
-                }
-                u32::from(next_byte - b'0')
-            }
-            b'a'..=b'z' => {
-                if next_byte > max_alpha_lower {
+    let data: u128 = data64 as u128;
+
+    if !bytes.is_empty() {
+        return handle_full_128_radix::<RADIX>(data, bytes, scale, passed_decimal_point, negative);
+    } else if !has_digit {
+        return Err(Error::from("Invalid decimal: no digits found"));
+    }
+
+    handle_data_radix(data, scale, negative)
+}
+
+const fn midpoint_for_radix(radix: u8) -> u8 {
+    if radix & 0x1 == 1 {
+        radix / 2
+    } else {
+        (radix + 1) / 2
+    }
+}
+
+#[inline(never)]
+#[cold]
+fn handle_full_128_radix<const RADIX: u8>(
+    mut data: u128,
+    mut bytes: &[u8],
+    mut scale: u8,
+    mut passed_decimal_point: bool,
+    negative: bool,
+) -> Result<Decimal, crate::Error> {
+    while !bytes.is_empty() {
+        let b = bytes[0];
+        match b {
+            b'0'..=b'9' | b'a'..=b'z' | b'A'..=b'Z' => {
+                let digit = match b {
+                    b'0'..=b'9' => u32::from(b - b'0'),
+                    b'a'..=b'z' if RADIX > 10 => u32::from(b - b'a' + 10),
+                    b'A'..=b'Z' if RADIX > 10 => u32::from(b - b'A' + 10),
+                    _ => return Err(Error::from("Invalid decimal: invalid character")),
+                };
+                if digit >= RADIX as u32 {
```

Comment:
same up here - do the duplicate checks for 10 have much performance impact?

Reply:
Interestingly, no difference on my Ryzen. I tried to get rid of it, but I am stuck by `use of generic parameter from outer function` where I cannot actually do something like `const MAX_DIGIT: u8 = RADIX + b'0';`